### PR TITLE
Piro Thyra Sensitivities

### PR DESCRIPTION
**Pull request by mperego**
Piro: Implemet adjoint sensitivities for Piro::NoxSolver (Thyra stack). Issue #3088

The adjoint system is solved by implicitly transposing the Jacobian operator and its preconditioner operator (thanks to @bartlettroscoe for the helping me with this).
This contains also a fix for update logic in ROL Thyra adapter, as well as a fix to allow Belos Thyra adapter to correctly select left/right preconditioner by @bartlettroscoe .

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
@mperego, is this ready to review and merge?

**Comment by mperego** [comment]
@bartlettroscoe Yes it is ready for review.

File: packages/piro/test/Main_AnalysisDriver_Tpetra.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      // END Builder
+
+      // Call the analysis routine
+      RCP<Thyra::VectorBase<double>> p;
+      status = Piro::PerformAnalysis(*piro, analysisParams, p);
+
+      if (Teuchos::nonnull(p)) {
+        // Can post-process results here
+        if (Proc==0) {
+          std::cout << "\nPiro_AnalysisDrvier:  Optimum printed above has exact soln = {1,3}" << std::endl;
```
Comment (rppawlo) [inline packages/piro/test/Main_AnalysisDriver_Tpetra.cpp]: Are the values of the optimum checked somewhere in this test? Why just printed to screen here? Could we change this driver to fail the test if the values aren't correct?

File: packages/piro/test/MockModelEval_A_Tpetra.hpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+typedef Tpetra::Map<int,int>  Tpetra_Map;
+typedef Tpetra::Vector<double,int,int>  Tpetra_Vector;
+typedef Tpetra::MultiVector<double,int,int>  Tpetra_MultiVector;
+typedef Tpetra::Operator<double,int,int>  Tpetra_Operator;
+typedef Tpetra::CrsGraph<int,int>  Tpetra_CrsGraph;
+typedef Tpetra::CrsMatrix<double,int,int>  Tpetra_CrsMatrix;
+typedef Thyra::TpetraOperatorVectorExtraction<
+    double, int, int> ConverterT;
+
+/** \brief Epetra-based Model Evaluator subclass for Charon!
```
Comment (rppawlo) [inline packages/piro/test/MockModelEval_A_Tpetra.hpp]: Why the Comment about Charon? Should probably be removed.

File: packages/piro/src/Piro_NOXSolver.hpp [hunk 3]
```diff
@@ -97,6 +97,11 @@ class NOXSolver
 
   bool exitUponFailedNOXSolve; 
 
+  enum DerivativeLayout { OP, COL, ROW };
```
Comment (rppawlo) [inline packages/piro/src/Piro_NOXSolver.hpp]: Can you add doxygen comments to explain the options for DerivativeLayout?

File: packages/piro/test/MockModelEval_A_Tpetra.hpp (hunk 2 above)
Comment (mperego) [reply packages/piro/test/MockModelEval_A_Tpetra.hpp]: Thanks, it's a copy and paste leftover. I'll get rid of it.

File: packages/piro/test/Main_AnalysisDriver_Tpetra.cpp (hunk 1 above)
Comment (mperego) [reply packages/piro/test/Main_AnalysisDriver_Tpetra.cpp]: This makes sense. I ported this driver from the Epetra version one, but it makes sense to have an error check. I'll put the error check in the Epetra driver as well.

File: packages/piro/src/Piro_NOXSolver.hpp (hunk 3 above)
Comment (mperego) [reply packages/piro/src/Piro_NOXSolver.hpp]: Sure. thanks.

**Comment by mperego** [comment]
@rppawlo, I have rebased my branch to make it  cleaner but unfortunately this got rid of all your comments. Sorry about that. Anyway, I implemented your suggestiond. Also I fixed the way I was creating the Tpetra map in the Tpetra driver. Thanks for your comments!

**Comment by mperego** [comment]
@rppawlo when you get a chance, can you re-approve this? Thanks, and sorry for the rebase.

**Comment by bartlettroscoe** [comment]
@trilinos/framework,

FYI: This PR branch was not merged.  Instead, it appears that the 5 commits were rebased and pushed.  My push logger recorded the push:

```
commit a2e4adbd24235a996571f808bd2f53f5440b6bb5
Author:     Mauro Perego <mperego@sandia.gov>
AuthorDate: Tue Jul 10 13:33:01 2018 -0600
Commit:     mperego <mperego@sandia.gov>
CommitDate: Fri Jul 13 08:41:50 2018 -0600

    Piro: Add tests to exercise the newly implemented adjoint sensitivities capability (issue #3088)

Commits pushed:
a2e4adb Piro: Add tests to exercise the newly implemented adjoint sensitivities capability (issue #3088)
704fa6f Piro: Add adjoint sensitivities capability for Thyra Piro_NOXSolver (issue #3088)       The adjoint system is solved by implicitly transposing the jacobian operator       and the preconditioner operator (thanks to @bartlettroscoe).
1f4ef6d Piro: allow for ROL trust region method in Piro::PerformAnalysis.
e407a34 ROL: Thyra Adapter, fix update logic (related to issue #3088)
00b55c4 Belos: Allow for explicit left and right preconditioners (related to issue #3087)
```

These commits are on the first-parent history of 'develop'.  Therefore, git bisection that hits these intermediate commits will not be robust because they may likely fail.

If we would just turn off the configuration option "Allow rebase merging" (see #2726) then we could avoid this happening in the future.

For example, try to click the "Revet" button above.  GitHub tells you you can't do it.  

On the other hand, a PR branch that is merged with merge commit like #3100 can be reverted (just click the button).

**Pull request merged**